File: small_problems/medium2/tri-angles.rb
# Triangles
# A triangle is classified as follows:
# - Right: One angle is a right angle (90 degrees)
# - Acute: All 3 angles are less than 90 degrees
# - Obtuse: One angle is greater than 90 degrees
# Other triangle rules:
# - Sum of the angles must be exactly 180 degrees
# - All angles must be greater than 0

# Write a method that takes 3 integers and returns a symbol:
# :right, :acute, :obtuse, or :invalid.
# The integers represent the angles of the triangle, in degrees
# - You may assume the angles are integer values (no float errors)

# Input: 3 integers
# - represent the angles of the 'triangle'
# Output: A symbol 
# - represent the type of triangle (acute/right/obtuse/invalid)

# Examples:
# triangle(60, 70, 50) == :acute
# triangle(30, 90, 60) == :right
# triangle(120, 50, 10) == :obtuse
# triangle(0, 90, 90) == :invalid
# triangle(50, 50, 50) == :invalid

# Data:
# - The 3 integer angles given as input
# - Probably store them in an array
# - Check if the triangle is valid first
# - If valid, then check which type it is

# Algorithm:
# Given 3 integers as input, angle1..angle3:
# Create an array to store the angles, angles = [...]
# Check if the angles form a valid triangle
#   - All angles are greater than 0, AND the sum of all angles must be 
#     exactly 180 
#   Return :invalid if triangle isn't valid
# If triangle is valid, calculate what type of triangle it is:
#   If all angles are less than 90 -> acute
#   If any angle is 90 -> right
#   If any angle is greater than 90 -> obtuse
#   - Check the value of the greatest angle
#     - If less than 90, acute; if 90, right; if more than 90, obtuse

def triangle(angle1, angle2, angle3)
  angles = [angle1, angle2, angle3]
  return :invalid unless valid_triangle?(angles)
  
  largest_angle = angles.max
  return :right if largest_angle == 90
  largest_angle > 90 ? :obtuse : :acute 
end

def valid_triangle?(angles)
  angles.size == 3 && angles.all?(&:positive?) &&
    angles.sum == 180
end

p triangle(60, 70, 50) == :acute
p triangle(30, 90, 60) == :right
p triangle(120, 50, 10) == :obtuse
p triangle(0, 90, 90) == :invalid
p triangle(50, 50, 50) == :invalid

File: javaScriptExercises/medium/tri-angles.js
// Triangles can be classified according to their internal angles as follows:
// 1) Right: One angle is 90 degrees exactly
// 2) Acute: All 3 angles are < 90 degrees
// 3) Obtuse: One angle is greater than 90 degrees

// For a triangle to be considered valid, the angles must sum to exactly 180 
// degrees, and every angle must be greater than 0.

// Write a function that takes the 3 angles of a triangle and returns one of 
// four strings: right, acute, obtuse, or invalid.

// Rules/Patterns
// - Assume that all angles have integer values, and are represented in degrees.

// Data: 
// > Input: 3 numbers (representing the angles of the triangle in degrees)
// 1) Validate the triangle
//   > A triangle is considered valid if:
//     a) All angles are > 0
//     b) The sum of all angles is exactly 180
// => 'invalid' if triangle is not valid

// 2) Classify the triangle
//   > Look at the largest given angle
//   - If maxAngle > 90   => obtuse
//   - If maxAngle === 90 => right
//   - If maxAngle < 90   => acute

// > Output: String (represents the classification of the triangle)

function triangle(angleA, angleB, angleC) {
  let angles = [ angleA, angleB, angleC ];

  if (!isValidTriangle(angles)) return 'invalid';

  let maxAngle = Math.max(...angles);
  if (maxAngle > 90) {
    return 'obtuse';
  } else if (maxAngle < 90) {
    return 'acute';
  } else { 
    return 'right';
  }
}

function isValidTriangle(angles) {
  return angles.length === 3 &&
         angles.every(angle => angle > 0) &&
         angles.reduce((sum, angle) => sum + angle, 0) === 180;
}

// Examples

console.log(triangle(60, 70, 50));       // "acute"
// [50, 60, 70]
console.log(triangle(30, 90, 60));       // "right"
// [30, 60, 90]
console.log(triangle(120, 50, 10));      // "obtuse"
// [10, 50, 120]
console.log(triangle(0, 90, 90));        // "invalid" (0 side)
console.log(triangle(50, 50, 50));       // "invalid" (sum !== 180)